sympy/stats/stochastic_process_types.py

```diff
@@ -883,21 +897,106 @@ def is_absorbing_chain(self):
+    def stationary_distribution(self, condition_set=False) -> ImmutableMatrix:
+        """
+        The stationary distribution is any row vector, p, that solves p = pP,
+        is row stochastic and each element in p must be nonnegative.
+        That means in matrix form: :math:`(P-I)^T p^T = 0` and
+        :math:`(1, ..., 1) p = 1`
+        where ``P`` is the one-step transition matrix.
+
+        All time-homogeneous Markov Chains with a finite state space
+        have at least one stationary distribution. In addition, if
+        a finite time-homogeneous Markov Chain is irreducible, the
+        stationary distribution is unique.
+
+        Parameters
+        ==========
+
+        condition_set : bool
+            If the chain has a symbolic size or transition matrix,
+            it will return a ``Lambda`` if ``False`` and return a
+            ``ConditionSet`` if ``True``.
+
+        Examples
+        ========
+
+        >>> from sympy.stats import DiscreteMarkovChain
+        >>> from sympy import Matrix, S
+
+        An irreducible Markov Chain
+
+        >>> T = Matrix([[S(1)/2, S(1)/2, 0],
+        ...             [S(4)/5, S(1)/5, 0],
+        ...             [1, 0, 0]])
+        >>> X = DiscreteMarkovChain('X', trans_probs=T)
+        >>> X.stationary_distribution()
+        Matrix([[8/13, 5/13, 0]])
+
+        A reducible Markov Chain
+
+        >>> T = Matrix([[S(1)/2, S(1)/2, 0],
+        ...             [S(4)/5, S(1)/5, 0],
+        ...             [0, 0, 1]])
+        >>> X = DiscreteMarkovChain('X', trans_probs=T)
+        >>> X.stationary_distribution()
+        Matrix([[8/13 - 8*tau0/13, 5/13 - 5*tau0/13, tau0]])
+
+        >>> Y = DiscreteMarkovChain('Y')
+        >>> Y.stationary_distribution()
+        Lambda((wm, _T), Eq(wm*_T, wm))
+
+        >>> Y.stationary_distribution(condition_set=True)
+        ConditionSet(wm, Eq(wm*_T, wm))
+
+        References
+        ==========
+
+        .. [1] https://www.probabilitycourse.com/chapter11/11_2_6_stationary_and_limiting_distributions.php
+        .. [2] https://galton.uchicago.edu/~yibi/teaching/stat317/2014/Lectures/Lecture4_6up.pdf
+
+        See Also
+        ========
+
+        sympy.stats.stochastic_process_types.DiscreteMarkovChain.limiting_distribution
+        """
         trans_probs = self.transition_probabilities
-        if trans_probs == None:
-            return None
-        if isinstance(trans_probs, MatrixSymbol):
-            wm = MatrixSymbol('wm', 1, trans_probs.shape[0])
-            return Lambda((wm, trans_probs), Eq(wm*trans_probs, wm))
-        w = IndexedBase('w')
-        wi = [w[i] for i in range(trans_probs.shape[0])]
-        wm = Matrix([wi])
-        eqs = (wm*trans_probs - wm).tolist()[0]
-        eqs.append(sum(wi) - 1)
-        soln = list(linsolve(eqs, wi))[0]
+        n = self.number_of_states
+
+        if n == 0:
+            return ImmutableMatrix(Matrix([[]]))
+
+        # symbolic matrix version
+        if not self._is_numeric:
```

Comment:
```suggestion
        if not isinstance(self.number_of_states, Integer) or isinstance(self.args[2], MatrixSymbol):
```

Reply:
I think irrespective of the fact, whether number of states is an integer or not, we do need a `Matrix` and not a `MatrixSymbol` to get our task done.

Reply:
`FunctionMatrix` can be of symbolic size but it does not inherit from `MatrixSymbol`. There are probably similar kinds of matrices with this kind of property. I think I'll change the statement to 
```python
if not isnstance(self.number_of_states, Integer) or isinstance(self.args[2], MatrixSymbol). 
```

Reply:
Ah I see. What about `if isnstance(self.args[2], `Matrix`)` as if it isn't Matrix we cannot go to the numeric version that you written after this `if` block? Do we really need to worry about `FunctionMatrix`? I see that there is a method, `as_explicit` which converts the `FunctionMatrix` to a concrete `Matrix`. So the user can convert `FunctionMatrix` to `Matrix` before creating their `DiscreteMarkovChain` object. May be you can remove that case, as `FunctionMatrix` won't be of much use as transition matrix since matrix multiplication require the whole matrix and matrix multiplication pretty much are a part of these chains.

Reply:
I think the transition probabilities should support any type of matrix to a reasonable extent. So any child of `MatrixBase` or `MatrixExpr` should hopefully be catered for. It is sadly not possible to obtain an explicit form of a symbolically sized matrix. I can add support for any matrix multiplied by a `FunctionMatrix` in the future. I'm surprised it hasn't been done yet.

Reply:
That will invite more confusion and more complications in the code. If it can be done with minimal changes then it will be fine to have such flexibility.

Reply:
Possibly a useful example of a matrix class for Markov chains would be a block matrix:
```julia
In [17]: M = MatrixSymbol('M', 2, 2)                                                                                                           

In [18]: B = BlockMatrix([[eye(2), M], [zeros(2, 2), eye(2)]])                                                                                 

In [19]: B                                                                                                                                     
Out[19]: 
⎡⎡1  0⎤        ⎤
⎢⎢    ⎥    M   ⎥
⎢⎣0  1⎦        ⎥
⎢              ⎥
⎢⎡0  0⎤  ⎡1  0⎤⎥
⎢⎢    ⎥  ⎢    ⎥⎥
⎣⎣0  0⎦  ⎣0  1⎦⎦

In [20]: B.as_explicit()                                                                                                                       
Out[20]: 
⎡1  0  M₀₀  M₀₁⎤
⎢              ⎥
⎢0  1  M₁₀  M₁₁⎥
⎢              ⎥
⎢0  0   1    0 ⎥
⎢              ⎥
⎣0  0   0    1 ⎦
```

Reply:
The explicit version works perfectly. But I realized now that `DiscreteMarkovChain` still only accepts `Matrix` and `MatrixSymbol`. It doesn't even accept `FunctionMatrix` even though I was testing for it :flushed:. 

I know @czgdp1807 [will want to keep it that way](https://github.com/sympy/sympy/pull/20026#discussion_r481686162) and I think @oscarbenjamin [suggested that we extend it to `MatrixBase` and `MatrixExpr`](https://github.com/sympy/sympy/pull/20026#discussion_r482284967) in my older PR.

So which one should we go with?

Reply:
I am still having the same reasoning. It's a bit risky and confusing to involve things which aren't in any way helping to improve performance or design. We can make any API as flexible as possible but that increases ambiguity too and then a time comes when it becomes too messy and causes a lot of effort to make even small changes. 
I am not suggesting to not make this flexible but the code is still in an early stage and rather on focusing on public APIs IMHO, we should focus on extensive testing of the implementation and ensure that it gives correct results for cases which are not there in the test files right now. 

In short, what I think is that APIs should be simple, less ambiguous and easy to use. Flexibility invites ambiguity, messy code which becomes too difficult to maintain. `MatrixBase` can be allowed since that would be quite easy to do and won't require much changes AFAICT but allowing something like, `FunctionMatrix` which is completely different and just delays the worst case that O(n^2) memory and o(n^2) time complexity won't be of much help. `MatrixExpr` may be added in distant future but at least not for now.